rfcs/0052-extensible-udt/0052-extensible-udt.md

```diff
@@ -0,0 +1,369 @@
+has a hidden `flags` field with all zeros.
+
+• If `flags & 0x1FFFFFFF` is 0x1, extension data will contain a
+[molecule](https://github.com/nervosnetwork/molecule) serialized `ScriptVec`
+structure:
+
+```
+table Script {
+    code_hash:      Byte32,
+    hash_type:      byte,
+    args:           Bytes,
+}
+
+vector ScriptVec <Script>
+```
+
+Each entry included in `ScriptVec` structure is interpreted as
+an extension script with additional behaviors. When an xUDT script is
+executed, it will run through each included extension script. Only when all
+extension scripts pass validation, will xUDT also consider the validation to be
+successful.
+
+An extension script can be loaded in any of the following ways:
+
+1. Some extension logics might have a predefined hash, for example, we can
+   use `0x0000 ... 0001` to represent regulation extension. The actual code for
+   such scripts can be embedded in xUDT script itself.
+2. If an input cell in the current transaction uses a lock script with the same
+   script hash as the current extension script, we can consider the extension
+   script to be validated already.
+3. If an extension script does not match any of the above criteria, xUDT will
+   use the code_hash and hash_type included in the extension script to
+   invoke [ckb_dlopen2](https://github.com/nervosnetwork/ckb-c-stdlib/blob/37eba3102100808ffc6fa2383bcf9e1e2651c8ea/ckb_dlfcn.h#L108-L113) function,
+   hoping to load a dynamically linked script from cell deps in the current
+   transaction. If a script can be located successfully, xUDT will then look for
+   an exported function with the following signature:
+
+```c
+int validate(int is_owner_mode, size_t extension_index, const uint8_t* args, size_t args_length);
+```
+
+`is_owner_mode` indicates if the current xUDT is unlocked via owner mode(as
+described by sUDT), `extension_index` refers to the index of the current
+extension in the `ScriptVec` structure. `args` and `args_length` are set to the
+script args included in `Script` structure of the current extension script.
+
+If this function returns 0, the current extension script validation is
+considered successful.
+
+• If `flags & 0x1FFFFFFF` is 0x2, extension data will contain the blake160 hash
+of the `ScriptVec` structure as explained in the previous section. The
+actual `raw_extension_data` (`ScriptVec`) structure data will be included in a
+witness field `input_type` or `output_type` contained in the current
+transaction. We will explain this part below. Choosing `input_type` or
+`output_type` depends on whether the type script is running on input or output
+cells. Under a lot of scenarios, it is `input_type`. But in the following example “Owner
+Mode Without Consuming Cell”, we can see it’s possible on `output_type`.
+
+### xUDT Witness
+
+The `input_type` or `output_type` field in witness has the following data
+structure in molecule format:
+
+```js
+table XudtWitnessInput {
```

Comment:
The name here is slightly confusing. While most cases would keep this data in `input_type` field, there are cases where it might be stored in `output_type` field. This means we probably should not name it `XXXInput`.

Reply:
Updating this name is easy. But the script code is freezed and the script is deployed.
Can we just add some comments about it's possible in `output_type` in some scenarios.

Reply:
I do understand the code is freezed, I'm not suggesting any change in behaviors.

But I do believe that API-side, we can tweak the name without any breaking code. There is even a chance the binary code would still stay the same even if we modify the field names in molecule source file, recreate molecule generated file, then recompile the C source file, assuming we've done sufficient purging. Since the RFC is still being written, I do recommend we spend enough time thinking about proper names.